src/App.tsx

```diff
@@ -1,26 +1,175 @@
+import { Todo } from './types/Todo';
+import { Status } from './types/Status';
+import { TodoFilter } from './components/TodoFilter';
+import { TodoItem } from './components/TodoItem';
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
-  }
-
-  return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">
-          React Todo App - Load Todos
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [newTodoTitle, setNewTodoTitle] = useState('');
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [errorMsg, setErrorMsg] = useState('');
+  const [filter, setFilter] = useState<Status>(Status.All);
+  const [processingTodos, setProcessingTodos] = useState<number[]>([]);
+
+  const addInputRef = useRef<HTMLInputElement>(null);
+  const allAreCompleted = todos.every(todo => todo.completed);
+  const hasTodos = todos.length > 0;
+
+  const filteredTodos = useMemo(
+    () => filterTodos(todos, filter),
+    [todos, filter],
+  );
+
+  const handleError = (error: string) => {
+    setErrorMsg(error);
+
+    setTimeout(() => {
+      setErrorMsg('');
+    }, 3000);
+  };
+
+  const onTodoAdd = (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+
+    if (!newTodoTitle.trim()) {
+      handleError('Title should not be empty');
+
+      return;
+    }
+
+    setTempTodo({
+      title: newTodoTitle.trim(),
+      id: 0,
+      userId: USER_ID,
+      completed: false,
+    });
+
+    addTodo(newTodoTitle.trim())
+      .then(newTodo => {
+        setTodos(prevTodos => [...prevTodos, newTodo]);
+        setNewTodoTitle('');
+      })
+      .catch(() => {
+        handleError('Unable to add a todo');
+        setTodos(todos);
+      })
+      .finally(() => {
+        setTempTodo(null);
+      });
+  };
+
+  const onTodoDelete = (todoId: number) => {
+    setProcessingTodos(prevProcessing => [...prevProcessing, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos(prev => prev.filter(prevTodo => prevTodo.id !== todoId));
+      })
+      .catch(() => handleError('Unable to delete a todo'))
+      .finally(() => {
+        setProcessingTodos(prevIds => prevIds.filter(id => id !== todoId));
+      });
+  };
+
+  const onTodoDeleteCompleted = () => {
+    const completedTodos = todos.filter(todo => todo.completed);
+
+    completedTodos.forEach(todo => onTodoDelete(todo.id));
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => handleError('Unable to load todos'));
+  }, []);
+
+  useEffect(() => {
+    addInputRef.current?.focus();
+  }, [todos, tempTodo]);
+
+  return !USER_ID ? (
+    <UserWarning />
```

Comment:
I guess you don't need `UserWarning` anymore

Reply:
> I guess you don't need `UserWarning` anymore

it's true, cuz it's a constant that is always there)